broker:  allow late-joining brokers, and execute rc1/rc3 on all broker ranks

garlick opened the pull request:
OK, this is a rework of PR #3057 which was withdrawn.  It changes broker startup and teardown to run rc1 and rc3 scripts on all ranks rather than just on rank 0.  The design is described [here](https://github.com/flux-framework/flux-core/blob/f0c585b2c76a6ef0b37c48368ebd5f580835afdc/src/broker/doc/README.md) and hasn't changed much since previous attempt.

Some differences:
* By default, all brokers must be online before the initial program is started.  This can be defeated by setting the `state-machine.quorum` attribute to the idset of a required set of brokers (e.g. systemd unit file might start broker with `-Sstate-machine.quorum=0`).
* There is no longer a published `shutdown` event.  Each broker watches its parent state transitions and entry to the SHUTDOWN state propagates across the instance that way.
* The shutdown class was removed and replaced with simpler code in state_machine.c.
* The hello class was removed and replaced with simpler "quorum" code in state_machine.c (resource module watches this for now)
* Timeout waiting for parent to respond in JOIN state was not implemented yet (deferred to future PR)
* Timeout waiting for children to disconnect in SHUTDOWN state was not implemented yet (deferred to future PR)

In this PR I was able to refactor the commits so that the switch to distributed rc1 and rc2 didn't need to be squashed together with a huge amount of change.  This was at the expense of including a small amount of incremental development in the PR, but I hope it's worth it to make the commits more readable.

TODO: tests

garlick commented:
Rebased on master to pickup python black PR (thanks@grondo!), and squashed LGTM fix.

For some reason, I'm seeing a bunch of failures on the centos 7 builder.  I'll try to track those down in a docker image.

garlick commented:
I've pushed some changes:
* Avoid situation that causes deadlock on centos 7 / zeromq-4.1.4 (must be a zeromq bug that was flxed later, but I haven't been able to find it in their changelogs)
* Speed up t0001-basic.t by tightening up flux-start killer timeout in one test, and suppressing rc1/rc3 some places where it doesn't seem to be a problem.
* Add timing spent in each state to state transition log entries
* Accept deprecated `--shutdown-grace` option and emit a warning, to avoid breaking sched tests

grondo commented:
Just started testing this branch on fluke (launching under Slurm).

I don't think this happened before (I can do more testing later):
```
 grondo@fluke108:~/git/flux-core.git$ srun -N4 --pty src/cmd/flux start flux getattr size
4
2020-09-03T23:51:55.712220Z resource.err[0]: hwloc-reload exited with rc=1
2020-09-03T23:51:55.713668Z sched-simple.err[0]: rpc_get (resource.acquire): Invalid argument
2020-09-03T23:51:55.713871Z sched-simple.crit[0]: fatal error: Invalid argument
```

Running `flux resource list` instead of `flux getattr size` avoids the errors (I'm guessing because resource list blocks until resources are available)

I think you may have mentioned this issue already somewhere. Apologies if so.

garlick commented:
Hmm, I think you found something there.   I guess the initial program must be completing before the resource module has finished its `flux hwloc reload`, so some ranks (starting with the leaves) begin to shut down,  then probably RPCs to those ranks from `flux hwloc reload` fail?  Hmm...

grondo commented:
It is pretty reproducible at 4 nodes.

Using `flux resource list` to time instance resource availability + normal shutdown, I think this branch is only slightly slower than v0.18.0 (the default on fluke)

```
 grondo@fluke108:~/git/flux-core.git$ time srun -N52 -n208 --pty flux start flux resource list
     STATE NNODES   NCORES    NGPUS
      free    208      208        0
 allocated      0        0        0
      down      0        0        0

real	0m13.182s
user	0m0.017s
sys	0m0.021s
 grondo@fluke108:~/git/flux-core.git$ time srun -N52 -n208 --pty src/cmd/flux start flux resource list
     STATE NNODES   NCORES    NGPUS
      free    208      208        0
 allocated      0        0        0
      down      0        0        0

real	0m16.100s
user	0m0.014s
sys	0m0.028s
```

Only 52 nodes are up on fluke right now so I can't go much bigger.

No other issues found! Nice work.

garlick commented:
Would it be too egregiously hacky to put `flux module list >/dev/null` in the cleanup scripts?  This would ensure the scheduler has completed its synchronous initialization before brokers start to come down.

For dynamic resource discovery longer term, we could gather hwloc data by loading a module on all ranks and having it push resources towards rank 0, so that gathering is complete by the time all the rc1's have completed?

grondo commented:
That doesn't seem too bad given the circumstances. Is there a way to put a comment in the cleanup script, so it is obvious why that line is there?

garlick commented:
Oh bummer, it looks like fluxion doesn't have `sched.resource-status` implemented yet so that won't work

garlick commented:
Another problem with fluxion is that it tries to unload `sched-simple` in rc1.  With the proposed changes, `sched-simple` is unresponsive until resource discovery is complete, which cannot be complete at that point (unless it is pre-populated in the KVS due to an instance restart).

dongahn commented:
> Oh bummer, it looks like fluxion doesn't have sched.resource-status implemented yet so that won't work

Hmmm. This should be there. What version are you using?

dongahn commented:
https://github.com/flux-framework/flux-sched/blob/master/qmanager/modules/qmanager.cpp#L456

dongahn commented:
> Another problem with fluxion is that it tries to unload sched-simple in rc1. With the proposed changes, sched-simple is unresponsive until resource discovery is complete, which cannot be complete at that point (unless the instance size=1, or it is pre-populated in the KVS due to an instance restart).

Maybe for testing, can we just don't load `sched-simple` during the initialization and see if fluxion can handle things correctly? Ultimately we may need a way for an external module to tell the system not to load certain services since it will replace them.

garlick commented:
> Hmmm. This should be there. What version are you using?

Thanks @dongahn, I am using the tip of master and can confirm the code is there.  Possibly some other race was causing the behavior I observed.

grondo commented:
Will #3177 help with sched-simple issue? I admit I haven't had a chance to look at that PR yet.

garlick commented:
> Will #3177 help with sched-simple issue? I admit I haven't had a chance to look at that PR yet.

Yeah, I think that would do it, although it also requires some changes to the fluxion test suite.  I think I was able to fix all the tests over there but one.

garlick commented:
> Maybe for testing, can we just don't load sched-simple during the initialization and see if fluxion can handle things correctly?

It's not just a testing problem, but yep that's the idea in #3177

dongahn commented:
> Thanks @dongahn, I am using the tip of master and can confirm the code is there. Possibly some other race was causing the behavior I observed.

One implementation difference between Fluxion and `sched-simple` is that `qmanager` will relay the request to `sched-fluxion-resource` via an additional RPC to get the resource list. Would that make any difference? Is `sched-fluxion-resource` loaded when `flux resource list` is issued?

garlick commented:
Yes, let me dig a little deeper and see if I can figure out what is going on.

Incidentally the request relay is synchronous.  That should probably be made asynchronous so qmanager doesn't become unresponsive in the event that fluxion-resource is slow to respond (presumably graph traversal can take a bit?)

garlick commented:
Oh I think the problem is:  some fluxion tests exit with the scheduler unloaded, then in the CLEANUP phase I get "no such service" when running `flux resource list`.  That makes sense.  In this PR I had already changed this makeshift synchronization to a KVS operation that doesn't involve the scheduler, so I think is resolved.  Sorry to bring it on your radar when I should have expected it.

dongahn commented:
> Incidentally the request relay is synchronous. That should probably be made asynchronous so qmanager doesn't become unresponsive in the event that fluxion-resource is slow to respond (presumably graph traversal can take a bit?)

Yeah, a lot of RPCs within in Fluxion is synchronous at this point.

I wanted to get a majority of features completed first before adding optimization.

I felt that there are still lots of churns like making startup nonblocking and etc which can affect correctness under async RPC, I haven't bother to do optimization just yet. Towards the winter time frame, this can be worked on. Won't be too difficult,  and I will also need to benchmark our baseline performance as part of that as well.

dongahn commented:
> Oh I think the problem is: some fluxion tests exit with the scheduler unloaded, then in the CLEANUP phase I get "no such service" when running flux resource list. That makes sense. In this PR I had already changed this makeshift synchronization to a KVS operation that doesn't involve the scheduler, so I think is resolved. Sorry to bring it on your radar when I should have expected it.

That makes sense.

BTW, I haven't followed exactly why we need `flux resource list`. But if you wanted to sync for a module to be loaded/available, I have used: `flux module stats`: https://github.com/flux-framework/flux-sched/blob/master/t/t1010-sync-modules.t#L29

garlick commented:
The issue is that we need to synchronize on the completion of resource discovery, otherwise we may start bringing down brokers before `flux hwloc reload` is finished sending RPCs to them all.  I can't use the generic service methods like ping or stats with the scheduler because CLEANUP doesn't know which scheduler is loaded and those methods are only valid for the service name == module name (e.g. sched-simple.stats not sched.stats).  Good thought though.

garlick commented:
This was rebased on top of #3177 and fluxion tests are all passing now with flux-framework/flux-sched#749 applied.

I made a couple of changes in the core `resource` module to abort a `resource.acquire` request and kill a running `flux hwloc reload` if the system starts to shut down before discovery completes.  I also fixed an error in the state machine implementation where an rc1 failure would skip over the SHUTDOWN state.  I failed to follow my own design document :-/

garlick commented:
Rebased on master after #3177 was merged.  Since flux-framework/flux-sched#749 was also merged, in theory the scheduler test should pass now.

garlick commented:
Hmm, asan builder seems to have timed out and the centos 7 travis builder hit some timeouts of this sort
```
flux-start: 0 (pid 2265) Killed
45631..46
4564ERROR: t0003-module.t - exited with status 137 (terminated by signal 9?)
```

grondo commented:
Ok tested out several runs on fluke (up to about 208 brokers) with no issues.

I also removed the `flux kvs get` synchronization in CLEANUP as a test and verified no hang, but instead see these errors:
```
2020-09-08T22:46:37.370836Z resource.err[0]: hwloc-reload killed by Killed
2020-09-08T22:46:37.378042Z sched-simple.err[0]: rpc_get (resource.acquire): No such file or directory
2020-09-08T22:46:37.379486Z sched-simple.crit[0]: fatal error: No such file or directory
```

These error messages could perhaps use some improvement. (Probably the sched-simple errors should print the future/rpc error message instead of blindly using `flux_log_error()`, sorry!)

garlick commented:
Thanks!

I noticed some spurious stderr on tests and realized I had messed up a couple of the test rc1/rc3 scripts, so pushed fixups for those.

I'll go ahead and remove that cleanup synchronization and fix the error reporting to be more informative.

garlick commented:
OK, fixups squashed, CLEANUP synchronization removed, and error message improved somewhat.  Hand editing your output to what I would expect after the change:
```
2020-09-08T22:46:37.370836Z resource.err[0]: hwloc-reload Killed
2020-09-08T22:46:37.378042Z sched-simple.err[0]: resource.acquire: resource discovery failed or interrupted
2020-09-08T22:46:37.379486Z sched-simple.crit[0]: fatal error: No such file or directory
```
For fun, I retested with flux-sched master and all its tests ran fine.

grondo commented on src/modules/sched-simple/sched.c:
```diff
@@ -501,7 +501,8 @@ static int ss_acquire_resources (flux_t *h, struct simple_sched *ss)
     ss->acquire_f = f;
     if (flux_rpc_get_unpack (f, "{s:o}", "resources", &o) < 0) {
-        flux_log_error (h, "rpc_get (resource.acquire)");
+        flux_log_error (h, "resource.acquire failed: %s",
+                       future_strerror (f, errno));
```
```suggestion
        flux_log (h, LOG_ERR, "resource.acquire failed: %s",
                  future_strerror (f, errno));
```

grondo reviewed (approved):
LGTM again!

Just one suggestion inline

garlick commented:
OK, I fixed that error message, and also added a couple more tests to the state machine sharness test covering failed rc1,rc2,rc3 (ensuring expected state transitions occur).

Since you approved I'll go ahead and set MWP.  This is a big change and there may be fallout, but given deadlines I think we should get it in and sort out any problems as we see them.

grondo commented:
Sounds good to me! Nice work!

garlick commented:
Thanks!
